src/sema/variables.rs

```diff
@@ -460,6 +461,143 @@ pub fn variable_decl<'a>(
                 ));
             }
 
+            let body;
+            let returns;
+
+            // If the return value of an accessor function is a struct, return the members rather than
+            // the struct. This is a particular inconsistency in Solidity which does not apply recursively,
+            // i.e. if the struct contains structs, then those values are returned as structs.
+            if let Type::Struct(str) = &ty {
+                let expr = Arc::new(expr);
+
+                if constant {
+                    let var_no = symtable
+                        .add(
+                            &pt::Identifier {
+                                loc: pt::Loc::Implicit,
+                                name: "_".into(),
+                            },
+                            ty.clone(),
+                            ns,
+                            VariableInitializer::Solidity(Some(expr.clone())),
+                            VariableUsage::LocalVariable,
+                            Some(pt::StorageLocation::Storage(pt::Loc::Implicit)),
+                        )
+                        .unwrap();
+
+                    symtable.vars[&var_no].read = true;
+                    symtable.vars[&var_no].assigned = true;
+
+                    let def = str.definition(ns);
+
+                    let list = def
+                        .fields
+                        .iter()
+                        .enumerate()
+                        .map(|(i, param)| Expression::Load {
+                            loc: pt::Loc::Implicit,
+                            ty: param.ty.clone(),
+                            expr: Expression::StructMember {
+                                loc: pt::Loc::Implicit,
+                                ty: Type::Ref(def.fields[i].ty.clone().into()),
+                                expr: Expression::Variable {
+                                    loc: pt::Loc::Implicit,
+                                    ty: ty.clone(),
+                                    var_no,
+                                }
+                                .into(),
+                                field: i,
+                            }
+                            .into(),
+                        })
+                        .collect();
+
+                    body = vec![
+                        Statement::VariableDecl(pt::Loc::Implicit, var_no, param, Some(expr)),
+                        Statement::Return(
+                            pt::Loc::Implicit,
+                            Some(Expression::List {
+                                loc: pt::Loc::Implicit,
+                                list,
+                            }),
+                        ),
+                    ];
```

Comment:
What do you think of building the function body during codegen?

Reply:
I can see the argument for keeping the ast more "pure", but on the other hand I don't think it is that ugly. However, I am not sure what this would look like, code wise.

We need to build the function prototype in sema, that includes the parameters and return values. Once we've calculated all of that, we've already done almost all the work to build up the body. We would need to do the same work again in codegen to build the body, or somehow share this code.

Maybe something for another day.

Reply:
In principle I agree with Lucas, it seems odd to sort of generate code not in _codegen_ but in sema. However there might be more such instances. I also thought a bit about it and didn't come to an immediate conclusion or solution to that question.